### PR TITLE
[java gen] - Support generated annotations

### LINKED ISSUE #4384: [java-gen] - Generated Java model should include new specific annotations (min, max, etc.)
### Is your enhancement related to a problem? Please describe

[Support for specific annotations](https://github.com/fabric8io/kubernetes-client/pull/4348) (e.g.: `@Max`, `@Min`, `@Pattern`, etc.), was added recently but the Java generator doesn't generate any of such annotations.

### Describe the solution you'd like

The Java generator should take the related keywords into account when they appear in CRDs (e.g.: `maximum`, `minimum`, `pattern`, etc.), and generate the corresponding annotations for them to appear in the Java model.

### Describe alternatives you've considered

N/A

### Additional context

See:
* [Introduce specific annotations for the generators](https://github.com/fabric8io/kubernetes-client/pull/4348) and specifically https://github.com/fabric8io/kubernetes-client/pull/4348#issuecomment-1236555434
* [Add support for Min, Max, Pattern and Nullable on CRD generator fields](https://github.com/fabric8io/kubernetes-client/pull/4298)

### PR DESCRIPTION
**Pull request by fabiobrz**
## Description

The changes are related to those in https://github.com/fabric8io/kubernetes-client/pull/4348 
The generator will now generate the expected annotations, i.e. those defined in `io.fabric8.generator.annotation`, the same ones that a reverse CRD generation from the Java model take into account.

Fix #4384 

## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [x] I Added [CHANGELOG](https://github.com/fabric8io/kubernetes-client/blob/master/CHANGELOG.md) entry regarding this change
 - [x] I have implemented unit tests to cover my changes
 - [x] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [x] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

I can suash the commits as soon as this is ready for merge or even earlier in case we agree on it. 
FYI @andreaTP, CC  @manusa

### CODE REVIEW

File: java-generator/core/pom.xml [hunk 1]
```diff
@@ -68,9 +68,8 @@
     </dependency>
 
     <dependency>
-      <groupId>javax.validation</groupId>
```
Comment (andreaTP) [inline java-generator/core/pom.xml:71]: Nice! would you mind removing the dependency from the integration tests(and everwhere under the `java-generator` folder) and in the relevant docs?

File: java-generator/core/src/main/java/io/fabric8/java/generator/nodes/AbstractJSONSchema2Pojo.java [hunk 2]
```diff
@@ -69,11 +81,17 @@ protected String getClassType() {
     return getType();
   }
 
-  protected AbstractJSONSchema2Pojo(Config config, String description, final boolean isNullable, JsonNode defaultValue) {
+  protected AbstractJSONSchema2Pojo(Config config, String description, final boolean isNullable, JsonNode defaultValue,
+      final Optional<ValidationProperties> validationProperties) {
```
Comment (manusa) [inline java-generator/core/src/main/java/io/fabric8/java/generator/nodes/AbstractJSONSchema2Pojo.java:85]: Please, never use Optionals as method parameters, use them only as return types.

The reasoning is very simple, consider I invoke the constructor with `super(theConfig, theDescription, theIsNullable, theDefaultalue, null)`, the NPE problem would still be there.

If you're willing to do this with an Optional, then the suggestion would be to keep the parameter as `ValidationProperties validationProperties`,
and then replace your constructor body with:
```java
Optional.ofNullable(values).ifPresent(values -> {
//...
```
Or better and simpler just use an if not null clause.

Comment (fabiobrz) [reply java-generator/core/src/main/java/io/fabric8/java/generator/nodes/AbstractJSONSchema2Pojo.java:85]: +1, thanks @manusa - fixing.

File: java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JObject.java [hunk 3]
```diff
@@ -229,6 +247,8 @@ public GeneratorResult generateJava() {
               new SingleMemberAnnotationExpr(
                   new Name("com.fasterxml.jackson.annotation.JsonSetter"),
                   new NameExpr("nulls = com.fasterxml.jackson.annotation.Nulls.SET")));
+          // let's also add Fabric8 specific `Nullable` annotation, which will be reflected on generated CRD
```
[the remark this answers is not in the document]
Comment (fabiobrz) [reply java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JObject.java:250]: Will do!

File: java-generator/core/src/main/java/io/fabric8/java/generator/nodes/AbstractJSONSchema2Pojo.java [hunk 4]
```diff
@@ -227,9 +255,74 @@ private static AbstractJSONSchema2Pojo fromJsonSchema(
...
+            prop.getDefault());
       default:
         throw new JavaGeneratorException("Unreachable " + nt.getType());
     }
   }
+
+  /**
+   * This stores the values for a generic JSON schema validation related properties, like `minimum`, `maximum`, etc.
+   */
+  public static class ValidationProperties {
```
[the remark this answers is not in the document]
Comment (fabiobrz) [reply java-generator/core/src/main/java/io/fabric8/java/generator/nodes/AbstractJSONSchema2Pojo.java:273]: Yeah, makes sense. I followed the CRD `AbstractJsonSchema::SchemaPropsOptions` class, but I have nothing against moving it out.

File: java-generator/core/pom.xml (hunk 1 above)
Comment (fabiobrz) [reply java-generator/core/pom.xml:71]: Nice catch, thanks. Will do.

**Comment by fabiobrz** [comment]
Hi @andreaTP and @manusa - latest changes pushed to address your review comments. Let me know if you spot anything else.

BTW - can I squash the commit now?

**Review by andreaTP — APPROVED** [review]
Thanks a lot for this contribution @fabiobrz !
LGTM!

**Comment by fabiobrz** [comment]
> Thanks a lot for this contribution @fabiobrz ! LGTM!

Welcome @andreaTP - squashing my commits then!

**Comment by fabiobrz** [comment]
Hi @manusa - the e2e tests failed after I squashed my commits, although it was green already before that.
I'll try to re-spin a run, but IMHO it should be an intermittent/random failure.

**Comment by andreaTP** [comment]
cc. @manusa looking forward for your input here 🙏

**Review by manusa — APPROVED** [review]
LGTM, thx!

I think it would be worth adding a note on the the breaking changes list, since some users who might be using the Java generator, rely on the javax validation annotations that are no longer added.

This also raises the question: `as a developer relying on the java generator, how can I further transform the generated code to replace fabric8-fabric8 specific annotations with annotations that might be usedeful for my project (e.g. javax.validation)?`

**Comment by andreaTP** [comment]
> worth adding a note on the the breaking changes list

I agree, just want to note that the Java-Generator is still in "preview", so breaking changes are fine there.
Maybe `6.2` is a good candidate to get the Java Generator out of the Preview?

Regarding your additional question, we should come up with something for sure ... 😓 I have a couple of half-baked ideas in mind but am not convinced yet, open to suggestions!

**Comment by fabiobrz** [comment]
> I think it would be worth adding a note on the the breaking changes list, since some users who might be using the Java generator, rely on the javax validation annotations that are no longer added.

Hi @manusa - thanks for your feedback, will do.

> This also raises the question: as a developer relying on the java generator, how can I further transform the generated code to replace fabric8-fabric8 specific annotations with annotations that might be usedeful for my project (e.g. javax.validation)?

I've been discussing this shortly with @andreaTP as well last week.
Yesterday I was thinking about it and I came to the conclusion that maybe we'd want to create one issue for supporting the generation of validation constraints annotations per each framework type.

This would allow for them to be worked on asynchronously and for the support to different framework to be released in steps.
Maybe we could start with supporting the `javax.validation.*` as a PoC, in order to evaluate whether the proposal is actually fitting.

WDYT? CC @andreaTP

**Comment by andreaTP** [comment]
> for supporting the generation of validation constraints annotations per each framework type

This is an interesting idea, but I'm afraid to add this complexity to the core `java-generator` project.

I'm thinking about adding the possibility to arbitrarily customize the generated code before emitting it, offering a Visitor or Traverse pattern that can perform generic post-processing.

On top of that mechanism, we can implement separate plugins for each of the features we want to support, and they can possibly live in separate repositories/projects.

Note: the described mechanism might be used as a foundation to implement optimizations, such as class deduplication and similar.

**Comment by fabiobrz** [comment]
> > for supporting the generation of validation constraints annotations per each framework type
> 
> This is an interesting idea, but I'm afraid to add this complexity to the core `java-generator` project.
> 
> I'm thinking about adding the possibility to arbitrarily customize the generated code before emitting it, offering a Visitor or Traverse pattern that can perform generic post-processing.
> 
> On top of that mechanism, we can implement separate plugins for each of the features we want to support, and they can possibly live in separate repositories/projects.
> 
> Note: the described mechanism might be used as a foundation to implement optimizations, such as class deduplication and similar.

Thanks @andreaTP! 
+1 on loading the implementation dynamically, i.e. prepare the foundation for plugins. It makes sense to me and I actually thought of it as well, in relation to the complexity increase for the `core` module.

**Comment by fabiobrz** [comment]
>  I think it would be worth adding a note on the the breaking changes list

Done, should be ready fr merge as soon as CI checks pass.

**Comment by andreaTP** [comment]
Thanks!

**Pull request merged**